### PR TITLE
Add a reshape_copy operator.

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88361
* #88063
* __->__ #88314
* #88329

The semantics is "as if" you did a reshape, but it always copied
even if the input was directly view'able.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -2984,3 +2984,7 @@
 
 - name: special_spherical_bessel_j0(Tensor x) -> Tensor
   x: non_differentiable
+
+- name: reshape_copy(Tensor self, SymInt[] size) -> Tensor
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:2988]: Would it work to make it a `copy: bool` kwarg like we did for `.to()` https://pytorch.org/docs/master/generated/torch.Tensor.to.html?highlight=#torch.Tensor.to ?

**Review by albanD — APPROVED** [review]
Just a question about consistency with .to()
But if you need an actual different function and the flag is not enough, the PR sounds good.

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (ezyang) [reply tools/autograd/derivatives.yaml:2988]: Even if I did add the kwarg, I would still have to add an extra operator, similar to how we have `_to_copy`

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -1574,6 +1574,23 @@ Tensor reshape_symint(const Tensor& self, c10::SymIntArrayRef proposed_shape) {
...
+  if (self.is_sparse()) {
+    TORCH_CHECK(0, "_reshape_copy is not implemented for sparse tensors");
+  }
+  c10::SymDimVector shape = infer_size_dv(proposed_shape, self.sym_numel());
+
+  if (self.is_mkldnn()) {
+    TORCH_CHECK(0, "_reshape_copy not implemented for mkldnn tesnors");
+  }
+
+  if (self.is_contiguous()) {
```
Comment (ngimel) [inline aten/src/ATen/native/TensorShape.cpp:1587]: do you need this conditional, you are making a clone anyway?

Comment (ezyang) [reply aten/src/ATen/native/TensorShape.cpp:1587]: The guts here don't really matter, the inside of this function is never traced.

Comment (ngimel) [reply aten/src/ATen/native/TensorShape.cpp:1587]: Unsafe view is safe really, it does check the sizes, it just doesn't set up autograd tracking that you don't need here.

Comment (ezyang) [reply aten/src/ATen/native/TensorShape.cpp:1587]: ok, I will update this

**Pull request closed without merging**